### DISCUSSION TITLE
How can I validate both the array and individual items using refine?

### DISCUSSION THREAD

**Discussion by tparvi** [colinhacks/zod#511] [Q&A]
Codesandbox: https://codesandbox.io/s/zod-refine-array-nh5b6

**Note**: this seems like the same issues as https://github.com/colinhacks/zod/issues/480

I want to validate the individual items inside the array and the whole array. So I want to make sure individual items is correct and the whole array is valid. There might be e.g. logic that item X cannot be in the array if it also contains item Y.

If you open the CSB and click Validate it runs the validation and you can see that `validateSingleValue` is called but `validateArray` is not called. My schema is as follows i.e. values is an array that has objects:

```
function validateArray(input: any): any {
  console.log("validateArray called with iput", input);
  return false;
}

function validateSingleValue(input: Value): any {
  console.log("validateSingleValue called with iput", input);
  return false;
}

const schema = z.object({
  languageVersions: z.record(
    z.object({
      values: z
        .array(
          z
            .object({
              value1: z.string(),
              value2: z.string()
            })
            .refine(validateSingleValue, "Value is invalid")
        )
        .refine(validateArray, "Array of values is invalid")
    })
  )
});
```

**Comment by harshagarwalnyu** [comment]
Heads up, superRefine won't save you here. A refinement only runs after the base schema parses, so with z.array(itemSchema).superRefine(...), if any item fails its own validation the array parse is already invalid and the superRefine never fires. I checked this on zod 3.25: feed it one bad item and the array-level superRefine doesn't run at all, you just get the item error.

So to report BOTH the per-item errors and an array-level rule in one pass, make the element schema permissive enough that the array always parses, then do the item checks and the array check yourself inside one superRefine:

  const schema = z.array(z.unknown()).superRefine((items, ctx) => {
    items.forEach((raw, i) => {
      const r = itemSchema.safeParse(raw);
      if (!r.success) {
        for (const issue of r.error.issues) {
          ctx.addIssue({ ...issue, path: [i, ...issue.path] });
        }
      }
    });

    // array-level rule, runs no matter what the items did
    const ids = items.map((x) => x?.id);
    if (new Set(ids).size !== ids.length) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: 'Duplicate IDs', path: [] });
    }
  });

Because z.unknown() never fails, the array always parses, so your superRefine always runs and you get item-level issues (with the right path) plus the array-level issue together. If you're fine with item errors short-circuiting the array rule, then plain z.array(itemSchema).superRefine(...) is fine as-is, just know the array check won't run until the items are valid.
